RecoBTag/ONNXRuntime/plugins/ParticleNetSonicJetTagsProducer.cc

```diff
@@ -0,0 +1,317 @@
+  preprocessParams.setAllowAnything();
+  preprocessParams.setComment("`preprocessParams` is deprecated, please use `preprocess_json` instead.");
+  desc.addOptional<edm::ParameterSetDescription>("preprocessParams", preprocessParams);
+  desc.add<std::vector<std::string>>("flav_names",
+                                     std::vector<std::string>{
+                                         "probTbcq",
+                                         "probTbqq",
+                                         "probTbc",
+                                         "probTbq",
+                                         "probWcq",
+                                         "probWqq",
+                                         "probZbb",
+                                         "probZcc",
+                                         "probZqq",
+                                         "probHbb",
+                                         "probHcc",
+                                         "probHqqqq",
+                                         "probQCDbb",
+                                         "probQCDcc",
+                                         "probQCDb",
+                                         "probQCDc",
+                                         "probQCDothers",
+                                     });
+  desc.addOptionalUntracked<bool>("debugMode", false);
+
+  descriptions.addWithDefaultLabel(desc);
+}
+
+void ParticleNetSonicJetTagsProducer::produce(edm::Event &iEvent,
+                                              const edm::EventSetup &iSetup,
+                                              Output const &iOutput) {
+  edm::Handle<TagInfoCollection> tag_infos;
+  iEvent.getByToken(src_, tag_infos);
+
+  // initialize output collection
+  std::vector<std::unique_ptr<JetTagCollection>> output_tags;
+  if (!tag_infos->empty()) {
+    auto jet_ref = tag_infos->begin()->jet();
+    auto ref2prod = edm::makeRefToBaseProdFrom(jet_ref, iEvent);
+    for (std::size_t i = 0; i < flav_names_.size(); i++) {
+      output_tags.emplace_back(std::make_unique<JetTagCollection>(ref2prod));
+    }
+  } else {
+    for (std::size_t i = 0; i < flav_names_.size(); i++) {
+      output_tags.emplace_back(std::make_unique<JetTagCollection>());
+    }
+  }
+  for (unsigned jet_n = 0; jet_n < tag_infos->size(); ++jet_n) {
+    const auto &taginfo = (*tag_infos)[jet_n];
+    std::vector<float> outputs(flav_names_.size(), 0);  // init as all zeros
+    if (!taginfo.features().empty() && jet_n == 0) {
+      // run prediction and get outputs
+      const auto &output1 = iOutput.begin()->second;
+      const auto &outputs_from_server = output1.fromServer<float>()[0];
+      std::copy(outputs_from_server.begin(), outputs_from_server.end(), outputs.begin());
+    }
+
+    const auto &jet_ref = tag_infos->at(jet_n).jet();
+    for (std::size_t flav_n = 0; flav_n < flav_names_.size(); flav_n++) {
+      (*(output_tags[flav_n]))[jet_ref] = outputs[flav_n];
+    }
```

Comment:
to avoid a copy here:
```cpp
auto filler = [&](const auto& outputs) {
  const auto &jet_ref = tag_infos->at(jet_n).jet();
  for (std::size_t flav_n = 0; flav_n < flav_names_.size(); flav_n++) {
    (*(output_tags[flav_n]))[jet_ref] = outputs[flav_n];
  }
};
if (!taginfo.features().empty() && jet_n == 0) {
  const auto &output1 = iOutput.begin()->second;
  const auto &outputs_from_server = output1.fromServer<float>()[0];
  filler(outputs_from_server);
}
else {
  filler(std::vector<float>(flav_names_.size(), 0));
}
```

Reply:
I removed the copying, but in the current version, I'm not using the lambda function so that I don't have to make the extra vector for the empty case (and I wasn't using a lambda when I was trying to debug the repeated fromServer call).  Just let me know if you'd like me to go with the lambda function

Reply:
It's probably fine without the lambda, since it's really only a few lines of code duplicated.